src/core/qgsmapsettings.cpp

```diff
@@ -400,6 +400,28 @@ QPolygonF QgsMapSettings::visiblePolygon() const
   return poly;
 }
 
+QPolygonF QgsMapSettings::visiblePolygonWithBuffer() const
+{
+  QPolygonF poly;
+
+  const QSize &sz = outputSize();
+  const QgsMapToPixel &m2p = mapToPixel();
+
+  // Transform tilebuffer in pixel.
+  // Original tilebuffer is in pixel and transformed only according
+  // extent width (see QgsWmsRenderContext::mapTileBuffer)
+
+  const double mapUnitsPerPixel = mExtent.width() / sz.width();
+  const double buffer = mExtentBuffer / mapUnitsPerPixel;
+
+  poly << m2p.toMapCoordinates( -buffer,                              -buffer ).toQPointF();
```

Comment:
can you fix the formatting here? (I realise it's matching QgsMapSettings::visiblePolygon(), but there's no need to propagate that)